### PR TITLE
[DLSP21] 10-1

### PR DESCRIPTION
**Pull request by malfusion**
Authors: Srishti Bhargava (sb7261), Jude Naveen Raj Ilango (jni215)

### CODE REVIEW

File: docs/en/week10/10-1.md [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+**What is missing from pretext tasks?**
+ 
+There is a fairly big mismatch between what is being solved in the pretext tasks and what needs to be achieved by the transfer tasks (classifying examples/detecting objects). 
+Thus, this pre-training is not suitable for final tasks. 
+
+One possible way to fix is to apply linear classifiers to the intermediate layers to figure out the features at each of the layers, once the pre-trained network is obtained. This way, performance at each layer can be assessed to figure out which layer's features to use for the transfer tasks. 
+
+Pre-trained features should satisfy two fundamental properties:
+1. Useful to represent how images relate to one another (Picture of 2 trees, should be able to figure out that they are the same concept)
+2. Be robust to nuisance factors (should be able to differentiate trees in different conditions - weather, across years, lighting, etc)
```
Comment (Atcold) [inline docs/en/week10/10-1.md:46]: Please, shrink. This is way too much text.

File: docs/en/week10/10-1.md [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+
+### Popular method for self-supervised learning
+In the recent years, a popular and common method for self supervised learning is to learn features that are robust to data augmentation. 
+Learn features such that, 
+
+<div>
+$$
+\begin{aligned}
+    f_{\theta}(I) = f_{\theta}(augment(I))
```
Comment (Atcold) [inline docs/en/week10/10-1.md:55]:
```suggestion
    f_{\theta}(I) = f_{\theta}(\text{augment}(I))
```
Here and elsewhere.

File: docs/en/week10/10-1.md [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+1. Maximize Similarity Objective (Contrastive Learning, Clustering, Distillation)
+2. Redundancy Reduction Objective
+
+In order to evaluate these networks, a subset of the Imagenet dataset (1.3 million images over 1000 categories) without labels is used. And pre-training is performed using a randomly initialized Resnet-50 model.
+
+**What are the different ways to perform transfer learning and evaluate its performance?**
+
+We can do one of 2 things:
+1. Train a linear classifier on top of the fixed features output by the pretrained network. This may be suitable for a classification task.
+2. Finetune the entire pretrained network on the target task. This may be more suitable for a detection task.
```
Comment (Atcold) [inline docs/en/week10/10-1.md:83]: Please, shrink.

File: docs/en/week10/10-1.md [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+A **Memory bank** can be used to maintain a momentum of activations across all features. Hence, a set of 1000 examples, would mean a memory bank of 1000 features. Every time an image undergoes a forward pass through the network, the memory bank is updated using the new embeddings. These memory bank features can then be used as negatives for contrastive loss. 
+
+#### Advantages of Memory bank:
+- It is compute efficient as it requires only one Forward Pass to compute the embeddings to store.
+
+#### Drawbacks of Memory bank:
+- Not an online solution, features are only stored once per epoch and hence becomes stale very quickly.
+- Needs a large amount of GPU RAM to store the features.
+- Hard to scale to millions of images, as storage also needs to scale to millions of features.
```
Comment (Atcold) [inline docs/en/week10/10-1.md:170]: Plz, shrink.

File: docs/en/week10/10-1.md [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+4. Faster and has smaller compute requirements than MoCo and SimCLR.
+5. Easy to train on a smaller number (4-8) of GPUs.
+
+
+## Pretraining on ImageNet without labels:
+Even though ImageNet without labels can be used for self supervised learning, there is an inherent bias in the dataset due to the curation/hand-selection process involved. ImageNet has the following properties due to curation:
+1. Images belong to 1000 specific classes
+2. Images contain a prominent object
+3. Images have very limited clutter and very few background concepts
+
```
Comment (Atcold) [inline docs/en/week10/10-1.md:260]: Plz, shrink.

**Review by Atcold — CHANGES_REQUESTED** [review]
Figures do not have captions.
All figures are 16 MB! Please, reduce the size to 1 MB in total. People are loading these pages from mobile phones…
Also, way too much text. Plz, summarise more.

**Comment by malfusion** [comment]
Added all the changes you had mentioned. Please review

File: docs/en/week10/10-1.md [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+- Compute efficient. Requires one Forward Pass to compute the embeddings.
+
+#### Drawbacks of Memory bank:
+- Not online. Features become stale very fast.
+- Requires large amount of GPU RAM.
+- Hard to scale storage to millions of images.
+
+## MoCo
+
+MoCo, or, Momentum Contrast, is a contrastive method that utilizes a memory bank to maintain a momentum of activations. It works by using 2 encoders. The $f_{\theta}$ encoder is the encoder that needs to be learnt. $f_{\theta_{EMA}}$ encoder maintains an exponential moving average of $f_{\theta}$.
```
Comment (Atcold) [inline docs/en/week10/10-1.md:166]: `EMA` should be encapsulated into `\text{}` here and everywhere else. It's test, it's not the product of three variables.

File: docs/en/week10/10-1.md [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+
+<center>
+<img src="{{site.baseurl}}/images/week10/10-1/contrastive-learning.png" style="width: 350px; background-color:#DCDCDC;" />
+<img src="{{site.baseurl}}/images/week10/10-1/clustering.png" style="width: 250px; background-color:#DCDCDC;" /><br>
+Figure 8: Difference between Contrastive learning (Left) and Clustering (Right)
+</center>
+
+
+## SwAV
+The SwAV (Swapping Assignments between Views) algorithm is an online clustering method. The idea here is to maximize the similarity of a given image $I$ and augmentation of that image $augment(I)$ and, in doing so, ensure they belong to the same cluster.
```
Comment (Atcold) [inline docs/en/week10/10-1.md:199]: Please, use `\text{}` for **ALL** non-standard function names, here and everywhere else.

File: docs/en/week10/10-1.md [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+Soft assignment solves this problem by representing a lot more classes (logically) than hard assignment. It gives a richer representation and is less sensitive to $k$ (number of classes) and hence $k/n$. However, if hard assignment is used, the class imbalance and the fixed value of $n/k$ will create inaccuracies.
+
+#### Training SwAV:
+
+<center>
+<img src="{{site.baseurl}}/images/week10/10-1/swav.png" style="width: 600px; background-color:#DCDCDC;" /><br>
+Figure 11: Architecture of SwAV
+</center>
+
+2 crops from an image are passed through the network to compute the corresponding codes. The $2^{nd}$ code is then predicted from the $1^{st}$ embedding and vice versa. This task forces the network to become invariant to data augmentations. Gradients are backpropagated onto the prototypes as well. Hence this model is updated in an online manner.
```
Comment (Atcold) [inline docs/en/week10/10-1.md:245]: `nd` is the product of `n` and `d`. Use `\text{}` if necessary. Here and everywhere else.

File: docs/en/week10/10-1.md [hunk 9]
```diff
@@ -0,0 +1,271 @@
+---
+lang-ref: ch.10-1
+title: Self Supervised Learning in Computer Vision
+lecturer: Ishan Mishra
+authors: Srishti Bhargava, Jude Naveen Raj Ilango
+date: 16 May 2021
+---
+
+## Self Supervised Learning for Computer Vision
```
Comment (Atcold) [inline docs/en/week10/10-1.md:9]:
```suggestion
```

Remove duplicate.

**Review by Atcold — CHANGES_REQUESTED** [review]
Very good, almost there.

**Comment by malfusion** [comment]
Added the modifications

**Review by Atcold — APPROVED** [review]
Great, thanks!

**Pull request merged**